libcudacxx/include/cuda/__mdspan/host_device_accessor.h

```diff
@@ -0,0 +1,283 @@
+#  pragma system_header
+#endif // no system header
+
+#include <cuda_runtime_api.h>
+
+#include <cuda/std/__type_traits/is_pointer.h>
+#include <cuda/std/cassert>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+template <typename _Accessor>
+struct host_accessor;
+
+template <typename _Accessor>
+struct device_accessor;
+
+template <typename _Accessor>
+struct managed_accessor;
+
+/***********************************************************************************************************************
+ * Host/Device/Managed Accessor Traits
+ **********************************************************************************************************************/
+
+template <typename T>
+inline constexpr bool is_host_accessor_v = false;
+
+template <typename T>
+inline constexpr bool is_device_accessor_v = false;
+
+template <typename T>
+inline constexpr bool is_managed_accessor_v = false;
+
+template <typename _Accessor>
+inline constexpr bool is_host_accessor_v<host_accessor<_Accessor>> = true;
+
+template <typename _Accessor>
+inline constexpr bool is_device_accessor_v<device_accessor<_Accessor>> = true;
+
+template <typename _Accessor>
+inline constexpr bool is_managed_accessor_v<managed_accessor<_Accessor>> = true;
+
+template <typename T>
+inline constexpr bool is_host_device_managed_accessor_v =
+  is_host_accessor_v<T> || is_device_accessor_v<T> || is_managed_accessor_v<T>;
+
+template <typename __data_handle_type>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr bool __is_device_pointer(__data_handle_type __p)
+{
+  return ::cuda::std::is_pointer_v<__data_handle_type>
+      && (__p == nullptr || __isGlobal(__p) || __isShared(__p) || __isConstant(__p) || __isGridConstant(__p)
+          || __isLocal(__p));
+}
+
+/***********************************************************************************************************************
+ * Host Accessor
+ **********************************************************************************************************************/
+
+template <typename _Accessor>
+struct host_accessor : public _Accessor
+{
+private:
+  using __data_handle_type = typename _Accessor::data_handle_type;
+  using __reference        = typename _Accessor::reference;
+
+  static constexpr bool __is_ctor_noexcept      = noexcept(_Accessor{});
+  static constexpr bool __is_copy_ctor_noexcept = noexcept(_Accessor{});
+  static constexpr bool __is_access_noexcept    = noexcept(_Accessor{}.access(__data_handle_type{}, 0));
+  static constexpr bool __is_offset_noexcept    = noexcept(_Accessor{}.offset(__data_handle_type{}, 0));
+
+  static_assert(!is_host_device_managed_accessor_v<_Accessor>,
+                "cuda::host_accessor/cuda::device_accessor/cuda::managed_accessor accessor cannot be nested");
+
+  template <typename __data_handle_type>
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_host_pointer(__data_handle_type __p)
+  {
+    if constexpr (::cuda::std::is_pointer_v<__data_handle_type>)
+    {
+      cudaPointerAttributes __attrib;
```

Comment:
This conflates the allocation type with locality.

Consider some application which has overriden `std::new` to back all allocation with cudaMallocManaged. In those cases you would fail since the memory type is `cudaMemoryTypeManaged`.

What you want is to confirm the validity of the `hostPointer` or the type is unregistered:
```
    return (pointer_attr.hostPointer != nullptr ||
            pointer_attr.type == cudaMemoryTypeUnregistered);
```